src/engine/strat_engine/backstore/crypt.rs

```diff
@@ -461,14 +479,96 @@ impl CryptHandle {
         &self.key_description
     }
 
+    /// Get the keyslot associated with the given token ID.
+    pub fn keyslots(&mut self, token_id: c_uint) -> Result<Option<Vec<c_uint>>> {
+        get_keyslot_number(&mut self.device, token_id)
+    }
+
+    /// Get info for the clevis binding.
+    pub fn clevis_info(&mut self) -> Result<Option<(String, Value)>> {
+        let json = match self
+            .device
+            .token_handle()
+            .json_get(CLEVIS_LUKS_TOKEN_ID)
+            .ok()
+        {
+            Some(j) => j,
+            None => return Ok(None),
+        };
+        let json_b64 = match json
+            .get("jwe")
+            .and_then(|map| map.get("protected"))
+            .and_then(|string| string.as_str())
+        {
+            Some(s) => s.to_owned(),
+            None => return Ok(None),
+        };
+        let json_bytes = decode(json_b64).map_err(|e| LibcryptErr::Other(e.to_string()))?;
+
+        let subjson: Value = serde_json::from_slice(json_bytes.as_slice())
+            .map_err(|e| LibcryptErr::Other(e.to_string()))?;
+
+        pin_dispatch(&subjson).map(Some)
+    }
+
     /// Activate encrypted Stratis device using the name stored in the
     /// Stratis token
-    pub fn activate(&mut self) -> Result<()> {
-        activate_and_check_device_path(
-            &mut self.device,
-            &self.key_description,
-            &self.name.to_owned(),
-        )
+    pub fn activate(&mut self, unlock_method: UnlockMethod) -> Result<()> {
+        match unlock_method {
+            UnlockMethod::Keyring => activate_and_check_device_path(
+                &mut self.device,
+                &self.key_description,
+                &self.name.to_owned(),
+            ),
+            UnlockMethod::Clevis => clevis_luks_unlock(&self.physical_path, &self.name)
+                .map_err(|e| LibcryptErr::Other(e.to_string())),
+        }
+    }
+
+    /// Bind the given device using clevis.
+    /// The json argument may contain stratis and possibly pin specific
+    /// configuration which is not used by clevis and must be removed.
+    /// The configuration is not checked in any other way.
+    pub fn clevis_bind(&mut self, keyfile_path: &Path, pin: &str, json: &Value) -> Result<()> {
+        let mut json = json.clone();
+
+        let yes = if pin == "tang" {
+            let map = if let Some(m) = json.as_object_mut() {
+                m
+            } else {
+                return Err(LibcryptErr::Other(format!(
+                    "configuration for Clevis is is not in JSON object format: {}",
+                    json
+                )));
+            };
+
+            map.remove(CLEVIS_TANG_TRUST_URL).is_some()
```

Comment:
Given that the value for this is a boolean, wouldn't it make more sense to do `map.remove(CLEVIS_TANG_TRUST_URL).and_then(|b| b.as_bool()).unwrap_or(false)` or something similar just in case someone passes a `false` value? In that case, this code would still pass `-y` which could be problematic for API users.